src/analyses/base.ml

```diff
@@ -1381,6 +1381,11 @@ struct
     | Q.InvariantGlobal g ->
       let g: V.t = Obj.obj g in
       query_invariant_global ctx g
+    | Q.MayOverflow e ->  
+      IntDomain.local_no_overflow := true;
+      if M.tracing then M.trace "apron" "exp %a\n" d_exp e;
+      ignore(query_evalint (Analyses.ask_of_ctx ctx) ctx.global ctx.local e);
+      !IntDomain.local_no_overflow ;
```

Comment:
Did you find an example where the old no overflow handling in the relational domain caused an unsoundness in the analysis results?

Reply:
The old no overflow handling didn't cause any unsoundness in the analysis results because the flag was ignored (see [comment](https://github.com/goblint/analyzer/pull/1330/files#diff-f389b2fa9a7f2374023c34533d99d4ed44f8da04d6f46bec9a089ab9ee67b2bdL522)). However this overflow handling is not sufficient for the analysis, due to apron giving up when there is a cast from signed integer to unsigned integer and vice versa even when enabling ```--set sem.int.signed_overflow assume_none```. This may work for vla if there are just integer, but for malloc cil cast the expression to unsigned long.
![image](https://github.com/goblint/analyzer/assets/53515503/51866c0c-d666-46e6-baa3-9a18b5f06849)

Reply:
If we decide to not ignore the old overflow handling, we do get unsoundness in the analysis as one can see in [6b797ad](https://github.com/FungOliver/analyzer/commit/6b797ad49f867132d91cc89cfa7617397631a538). 
```
// PARAM: --enable ana.int.interval --set ana.activated[+] apron
#include <stdlib.h>
#include <goblint.h>

int main()
{
    unsigned int len = rand();
    len = len % 10;

    for (int i = 0; i -1 < len ; i++)
    {
        int tmp = i;
    }
}
```
This results in 
```
[Error][Analyzer][Unsound] both branches over condition '(unsigned int )(i - 1) < len' are dead (tests/regression/100-limitation/01-crash.c:9:9-9:37)
[Error][Analyzer][Unsound] Both branches dead
```

Reply:
I, unfortunately, also found out that the current overflow handling in this PR also suffers from this problem. The queries for EvalInt are cached, which results in the local_no_overflow not being set correctly. 

I have two ideas to solve this problem, which don't involve rewriting the whole evalint function. The first would be disabling the cache lookup with another global flag. The other would be recursively checking if the expression evaluates to Top, which is needed e.g. ``` ( (int) i < len ) && ( (int) j < len) ``` needs to be recursively checked.

Reply:
I implemented the idea from above. However, because evalint calls evalint in apron, the no_overflow function is called again. The no_overflow flag in the intdomain then gets reset, which leads to an imprecise calculation of the no_overflow flag. It seems like there is no other option, than rewriting the whole evalint in base to include the overflow information....

Reply:
> I, unfortunately, found out that the current overflow handling in this PR also suffers from this problem. The queries for EvalInt are cached, which results in the local_no_overflow not being set correctly.
> 
> One idea to solve this problem, which don't involve rewriting the whole evalint function would be to disable the cache lookup with another global flag.

Disabling the cache with a global flag seems to work. However, I agree that this solution may not be the prettiest. After having a look at the overflow handling in [ Linear two var equality #1297 ](https://github.com/goblint/analyzer/pull/1297) and testing a bit, it seems like their overflow handling may also work my analysis.